File: packages/engine/src/draft.ts
import type OpenAI from "openai";
import type { Block, BlueprintContent, BlueprintSection, DocSection } from "@runoff/core";
import { parseSectionText } from "./dialect.js";
import { MODEL, systemPrompt, sectionUserPrompt, type ScopedMemory } from "./prompts.js";

const tools = [
  {
    type: "function",
    function: {
      name: "ask_user",
      description:
        "Ask the user a blocking-with-fallback question when the source data or its business framing is genuinely ambiguous. Never ask about report mechanics (the dialect, citation markers, locator grammar) — the user never sees those; resolve them yourself and raise_flag if precision suffers. Continue drafting using the fallback unless an answer arrives.",
      strict: true,
      parameters: {
        type: "object",
        properties: {
          question: { type: "string" },
          options: { type: "array", items: { type: "string" } },
          fallback: { type: "string" },
          deadlineSection: { type: "string" },
        },
        required: ["question", "options", "fallback", "deadlineSection"],
        additionalProperties: false,
      },
    },
  },
  {
    type: "function",
    function: {
      name: "raise_flag",
      description:
        "Flag a passage that needs the user's judgment before the report can be released. Does not stop drafting.",
      strict: true,
      parameters: {
        type: "object",
        properties: {
          question: { type: "string" },
          options: { type: "array", items: { type: "string" } },
        },
        required: ["question", "options"],
        additionalProperties: false,
      },
    },
  },
] as const;

export interface DraftCallbacks {
  onDelta(text: string): void;
  onFlag(f: { question: string; options: string[] }): void;
  onQuestion(q: { question: string; options: string[]; fallback: string; deadlineSection: string }): void;
}

/**
 * Thrown when the model refuses to draft a section. The orchestrator catches
 * this per-section (emitting `section_failed`) so one refusal does not fail the
 * whole run; any other error propagates and fails the run.
 */
export class RefusalError extends Error {
  constructor(message = "model refused to draft this section") {
    super(message);
    this.name = "RefusalError";
  }
}

export interface DraftResult {
  raw: string;
  blocks: Block[];
}

/** An accumulated tool call, assembled from streamed `delta.tool_calls` chunks. */
interface AccumulatedToolCall {
  id: string;
  name: string;
  arguments: string;
}

export async function draftSection(opts: {
  client: OpenAI;
  content: BlueprintContent;
  section: BlueprintSection;
  dataBlock: string;
  completed: DocSection[];
  steers: string[];
  answers: { question: string; answer: string }[];
  retryFeedback?: string;
  previousSectionText?: string;
  memories?: ScopedMemory[];
  cb: DraftCallbacks;
}): Promise<DraftResult> {
  const { client, cb } = opts;
  const messages: any[] = [
    { role: "system", content: systemPrompt(opts.content, opts.memories) },
    { role: "user", content: sectionUserPrompt(opts) },
  ];
  let raw = "";
  let maxTokens = 16000;
  let lengthRetried = false;

  for (let iter = 0; iter < 6; iter++) {
    const stream = await (client as any).chat.completions.create({
      model: MODEL,
      stream: true,
      messages,
      tools,
      max_completion_tokens: maxTokens,
    });

    let turnText = "";
    let refusalText = "";
    let finishReason: string | null = null;
    const toolCalls: AccumulatedToolCall[] = [];

    for await (const chunk of stream) {
      const choice = chunk.choices?.[0];
      if (!choice) continue;
      const delta = choice.delta ?? {};

      if (typeof delta.content === "string" && delta.content.length) {
        turnText += delta.content;
        cb.onDelta(delta.content);
      }
      if (typeof delta.refusal === "string" && delta.refusal.length) {
        refusalText += delta.refusal;
      }
      if (Array.isArray(delta.tool_calls)) {
        for (const tc of delta.tool_calls) {
          const idx = tc.index ?? 0;
          const acc = (toolCalls[idx] ??= { id: "", name: "", arguments: "" });
          if (tc.id) acc.id = tc.id;
          if (tc.function?.name) acc.name = tc.function.name;
          if (tc.function?.arguments) acc.arguments += tc.function.arguments;
        }
      }
      if (choice.finish_reason) finishReason = choice.finish_reason;
    }

    raw = turnText;

    if (refusalText) throw new RefusalError();

    if (finishReason === "tool_calls") {
      const results: any[] = [];
      for (const call of toolCalls) {
        if (!call?.name) continue;
        // Guard the tool-argument parse: one malformed payload must not throw and
        // sink the whole run. Skip the callback and tell the model to continue.
        let parsed: any;
        try {
          parsed = JSON.parse(call.arguments || "{}");
        } catch {
          results.push({ role: "tool", tool_call_id: call.id, content: "Invalid tool arguments — ignored. Continue drafting." });
          continue;
        }
        let content: string;
        if (call.name === "ask_user") {
          cb.onQuestion(parsed);
          content =
            "The user will answer asynchronously. Proceed now using your stated fallback; an answer may be injected into later sections.";
        } else {
          cb.onFlag(parsed);
          content = "Flag recorded for the user's review. Continue drafting.";
        }
        results.push({ role: "tool", tool_call_id: call.id, content });
      }
      messages.push({
        role: "assistant",
        content: turnText || null,
        tool_calls: toolCalls
          .filter((c) => c?.name)
          .map((c) => ({ id: c.id, type: "function", function: { name: c.name, arguments: c.arguments } })),
      });
      for (const r of results) messages.push(r);
      continue;
    }

    // The model hit the token ceiling mid-draft. Retry the same turn once with a
    // larger budget; if it truncates again, accept the truncated text and let the
    // orchestrator proceed normally rather than failing the run.
    if (finishReason === "length" && !lengthRetried) {
      lengthRetried = true;
      maxTokens = 32000;
      continue;
    }

    break;
  }

  return { raw, blocks: parseSectionText(raw) };
}

File: backend/runoff_api/engine/draft.py
"""Port of packages/engine/src/draft.ts — the sync streaming section draft loop.

The LLM heart of the run pipeline: it builds the two-message request (system +
section user prompt from engine/prompts.py), streams the completion, assembles
tool-call fragments by index, dispatches ask_user/raise_flag through the
callbacks, retries once on a length cutoff, and parses the accumulated text into
dialect blocks. TS wins on every statement.

The request payload (model, messages, tools, sampling params) is field-for-field
identical to the TS ``create(...)`` object literal — a later task byte-checks it.

Runtime shapes are plain dicts with camelCase keys. ``cb`` is any object exposing
``on_delta(text)``, ``on_question(q)``, ``on_flag(f)`` (mirror DraftCallbacks).
"""

import json

from runoff_api.core.dialect import parse_section_text
from runoff_api.engine.prompts import MODEL, section_user_prompt, system_prompt

_TOOLS = [
    {
        "type": "function",
        "function": {
            "name": "ask_user",
            "description": (
                "Ask the user a blocking-with-fallback question when the source data or its "
                "business framing is genuinely ambiguous. Never ask about report mechanics (the "
                "dialect, citation markers, locator grammar) — the user never sees those; resolve "
                "them yourself and raise_flag if precision suffers. Continue drafting using the "
                "fallback unless an answer arrives."
            ),
            "strict": True,
            "parameters": {
                "type": "object",
                "properties": {
                    "question": {"type": "string"},
                    "options": {"type": "array", "items": {"type": "string"}},
                    "fallback": {"type": "string"},
                    "deadlineSection": {"type": "string"},
                },
                "required": ["question", "options", "fallback", "deadlineSection"],
                "additionalProperties": False,
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "raise_flag",
            "description": (
                "Flag a passage that needs the user's judgment before the report can be released. "
                "Does not stop drafting."
            ),
            "strict": True,
            "parameters": {
                "type": "object",
                "properties": {
                    "question": {"type": "string"},
                    "options": {"type": "array", "items": {"type": "string"}},
                },
                "required": ["question", "options"],
                "additionalProperties": False,
            },
        },
    },
]


class RefusalError(Exception):
    """Thrown when the model refuses to draft a section. The orchestrator catches
    this per-section (emitting ``section_failed``) so one refusal does not fail the
    whole run; any other error propagates and fails the run.
    """

    def __init__(self, message: str = "model refused to draft this section"):
        super().__init__(message)
        self.name = "RefusalError"


def draft_section(
    *,
    client,
    content: dict,
    section: dict,
    data_block: str,
    completed: list[dict],
    steers: list[str],
    answers: list[dict],
    retry_feedback: str | None = None,
    previous_section_text: str | None = None,
    memories: list[dict] | None = None,
    cb,
) -> dict:
    messages: list[dict] = [
        {"role": "system", "content": system_prompt(content, memories)},
        {
            "role": "user",
            "content": section_user_prompt(
                section=section,
                dataBlock=data_block,
                completed=completed,
                steers=steers,
                answers=answers,
                retryFeedback=retry_feedback,
                previousSectionText=previous_section_text,
            ),
        },
    ]
    raw = ""
    max_tokens = 16000
    length_retried = False

    for _iter in range(6):
        stream = client.chat.completions.create(
            model=MODEL,
            stream=True,
            messages=messages,
            tools=_TOOLS,
            max_completion_tokens=max_tokens,
        )

        turn_text = ""
        refusal_text = ""
        finish_reason = None
        tool_calls: dict[int, dict] = {}

        for chunk in stream:
            choices = getattr(chunk, "choices", None)
            choice = choices[0] if choices else None
            if not choice:
                continue
            delta = getattr(choice, "delta", None)

            content_delta = getattr(delta, "content", None)
            if isinstance(content_delta, str) and content_delta:
                turn_text += content_delta
                cb.on_delta(content_delta)
            refusal = getattr(delta, "refusal", None)
            if isinstance(refusal, str) and refusal:
                refusal_text += refusal
            deltas_tool_calls = getattr(delta, "tool_calls", None)
            if isinstance(deltas_tool_calls, list):
                for tc in deltas_tool_calls:
                    idx = getattr(tc, "index", None)
                    if idx is None:
                        idx = 0
                    acc = tool_calls.get(idx)
                    if acc is None:
                        acc = {"id": "", "name": "", "arguments": ""}
                        tool_calls[idx] = acc
                    if getattr(tc, "id", None):
                        acc["id"] = tc.id
                    fn = getattr(tc, "function", None)
                    if fn is not None:
                        if getattr(fn, "name", None):
                            acc["name"] = fn.name
                        if getattr(fn, "arguments", None):
                            acc["arguments"] += fn.arguments
            if getattr(choice, "finish_reason", None):
                finish_reason = choice.finish_reason

        raw = turn_text

        if refusal_text:
            raise RefusalError()

        if finish_reason == "tool_calls":
            results: list[dict] = []
            for call in tool_calls.values():
                if not call["name"]:
                    continue
                # Guard the tool-argument parse: one malformed payload must not throw and
                # sink the whole run. Skip the callback and tell the model to continue.
                try:
                    parsed = json.loads(call["arguments"] or "{}")
                except Exception:  # noqa: BLE001 — mirrors the TS catch-all
                    results.append({
                        "role": "tool",
                        "tool_call_id": call["id"],
                        "content": "Invalid tool arguments — ignored. Continue drafting.",
                    })
                    continue
                if call["name"] == "ask_user":
                    cb.on_question(parsed)
                    tool_content = (
                        "The user will answer asynchronously. Proceed now using your stated "
                        "fallback; an answer may be injected into later sections."
                    )
                else:
                    cb.on_flag(parsed)
                    tool_content = "Flag recorded for the user's review. Continue drafting."
                results.append({"role": "tool", "tool_call_id": call["id"], "content": tool_content})
            messages.append({
                "role": "assistant",
                "content": turn_text or None,
                "tool_calls": [
                    {
                        "id": c["id"],
                        "type": "function",
                        "function": {"name": c["name"], "arguments": c["arguments"]},
                    }
                    for c in tool_calls.values()
                    if c["name"]
                ],
            })
            for r in results:
                messages.append(r)
            continue

        # The model hit the token ceiling mid-draft. Retry the same turn once with a
        # larger budget; if it truncates again, accept the truncated text and let the
        # orchestrator proceed normally rather than failing the run.
        if finish_reason == "length" and not length_retried:
            length_retried = True
            max_tokens = 32000
            continue

        break

    return {"raw": raw, "blocks": parse_section_text(raw)}
